007-OFEP-flag-change-events (#25)
open-feature/ofep: 007-OFEP-flag-change-events.md

## 007-OFEP-flag-change-events

## State: DRAFTING

Some flag SDKs support listening for flag value changes or general configuration changes ([launchdarkly](https://docs.launchdarkly.com/sdk/features/flag-changes), [cloudbees](https://docs.cloudbees.com/docs/cloudbees-feature-management/latest/reporting/configuration-fetched-handler), flagd). This can allow us to use an event-based paradigm for consuming flags. Client apps may use feature flags for characteristics that aren't specifically tied to a user-action, making imperative flag evaluation a less-than-ideal solution. A server application could listen to a flag that changes some operational behavior.

Examples:

- A web client's banner color could be updated when an associated flag is changed, without any user action.
- A web service could subscribe to a flag that controls its log-level.

## Design

> The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a particular flag value or the configuration in general. When a _application author_ registers a handler, the provider would react by appropriately registering a listener on it's SDK, polling it's REST API, etc. If the flag (or configuration in general) is updated, the provider would call the registered handler. The handler might be called with the flag key, or perhaps some other metadata pertaining to the configuration update. **Flag values would not be provided to the handler**, the handler would simply run, indicating the associated configuration had changed. The _application author_ would then perform a flag evaluation for the changed flag. This is consistent with the aforementioned flag systems, which do not evaluate the flags when they've changed. One reason is that no dynamic context can be reasonably provided in the case of events, since the event is driven by a change in the flag management system, not a user-action.

Reviewer: ```suggestion
The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a particular flag value or the configuration in general. When an _application author_ registers a handler, the provider would react by appropriately registering a listener on it's SDK, polling it's REST API, etc. If the flag (or configuration in general) is updated, the provider would call the registered handler. The handler might be called with the flag key, or perhaps some other metadata pertaining to the configuration update. **Flag values would not be provided to the handler**, the handler would simply run, indicating the associated configuration had changed. The _application author_ would then perform a flag evaluation for the changed flag. This is consistent with the aforementioned flag systems, which do not evaluate the flags when they've changed. One reason is that no dynamic context can be reasonably provided in the case of events, since the event is driven by a change in the flag management system, not a user-action.
```

Per-flag design:

```ts
const client = OpenFeature.getClient();

// subscribe to changes in a string flag called "hex-color", and evaluate it when it's updated

> client.addHandler("hex-color", async (key) => {

Reviewer 2: +1 for this direct approach, looping through a list of changed flags seems unnecessarily cumbersome (other than to support vendors that don't offer this)

> client.addHandler("hex-color", async (key) => {

Reviewer 3: I like the look of this approach over the broad `configuration-change`.

May I also suggest the addition of a `client.addHandlerWithContext` method taking the `EvaluationContext` as an argument. In processes where the `EvaluationContext` is static for the full lifecycle of a process or client instance this will allow for an on change event to a specific flag. This would also reduce the need for `client.get{type}Value` calls at the top of each handler function.

```
client.addHandlerWithContext("hex-color", {"foo":"bar"}, (value) => {
     console.log(`hex-color value change: ${value}`);
})
```

Reviewer: I don't follow this. What does the context provide here? It doesn't seem to be passed into your resulting function..

Author: I like the idea of providing a context to the handler, but I don't think it will enable us to get the values with a change event, we'd still need to pass the context and call the evaluation in the handler. That's because most SDKs themselves don't allow you to pass the context to their handler, and don't provide flag values when the handler fires:

>The event data consists only of the flag key. It does not contain a flag value, because in server-side SDKs, flags only have values when they are evaluated for a specific set of user properties. (https://docs.launchdarkly.com/sdk/features/flag-changes)

This is LD doc, but it's not unusual. Cloudbees also doesn't provider values with their event.

Author: I've left this out of my current implementation. For now we won't pass flag values to the handlers either. They will have to be evaluated.

  console.log(
    `Got update for ${key}, new value is ${await client.getStringValue(
      key,
      "000000"
    )}`
  );
});
```

General configuration change design:

```ts
const client = OpenFeature.getClient();

// subscribe to changes in the configuration in general, and evaluate flagd it when it's updated
// this might be easier to implement across vendors, since many don't allow subscription to a particular flag, but general configuration changes

> client.on("configuration-change", async (changeMetadata) => {

Reviewer: Instead of having both methods, why don't we have changeMetadata just contain a list of flags which were updated if applicable?

Reviewer 4: Do we have an example of `changeMetadata`? I'm only familiar with the LaunchDarkly implementation of flag listeners that can listen for all changes, or a change given a `flag` and a `context`.

Author: > Do we have an example of `changeMetadata`?

I don't at this point, it was just food for thought. I somewhat doubt we could get any consistent information across different SDKs that could be useful there.

[Cloudbees has some metadata about the change](https://docs.cloudbees.com/docs/cloudbees-feature-management/latest/reporting/configuration-fetched-handler), but again, it seems like this would be widely different across SDKs, so we might want to leave out any sort of parameter here.

Reviewer 5: I like the ```configuration-change``` approach, but is there a half-way house, were we can define an optional filter.   I'm wondering if you had an application connected to a project with a few hundred flags, but only cared about 10, would you get alot of unnecessary interruptions ?

Author: I will add more detail. I didn't want to go too far if there wasn't much interest, but it seems like there is.

Author: I've made significant updates to this OFEP, addressing many of the concerns in this thread, hopefully.

Author: @weyert I've also updated the change event payload.

  console.log(
    `Got update for configuration, new value for hex-color is ${await client.getStringValue(
      key,
      "000000"
    )}`
  );
});
```

## Benefits

> Ability to use event-based flag evaluation paradigms.

Reviewer 6: Could use a few more sentences what the `event-based flag evaluation paradigms` are? Is this a common thing along feature flag systems? (first time I hear about it)

Author: I think I've addressed this.

## Caveats

> - Not all providers can reasonably implement this... some SDKs don't support subscriptions or configuration change notifications, for example. Should these providers simply never fire events?

Reviewer 6: I think it would make sense for this proposal to answer this question

Reviewer 7: Addressed or at least mitigated by this proposal.
https://github.com/open-feature/ofep/pull/36

- Implementation in providers will likely be more divergent than imperative evaluation (based on demo work below).
- Re-evaluating a flag after the handler fires sometimes means another round-trip, but most of the time in this case, things are locally cached - really dependant on SDK implementation. This is the pattern used by some vendors though, so it's not unprecedented.
- Some SDKs can't identify individual flags have changed, so all registered handlers must fire (see cloudbees provider demo). This might not be such a big deal since they are likely no-ops.

> - We may need a way to "shutdown" providers, cancelling all handlers.

Reviewer 6: I think we could add this to this proposal or raise a separate improvement proposal for this. I think this is important to have -- to ensure resources are getting correctly disposed off

Reviewer 7: Addressed by this proposal.
https://github.com/open-feature/ofep/pull/30

- We may need a way to "move" handlers from one provider to another if the provider is changed (recreate them in the new provider).

> - We should probably consider other events, like "SDK-READY" as well, which would fit well into design 2, above.

Reviewer 5: I think an event like SDK-READY would make alot of sense.   The current evaluation spec lets us provide a default value, which could be returned during an application startup before the SDK is ready, but there is no way for the consumer to know they got the default vs what they may have wanted.  Being able to listen of SDK-READY effectively lets them handle this, without blocking an application from starting up.

## 007-OFEP-flag-change-events

## State: DRAFTING

> Some flag SDKs support listening for flag value changes or general configuration changes ([launchdarkly](https://docs.launchdarkly.com/sdk/features/flag-changes), [cloudbees](https://docs.cloudbees.com/docs/cloudbees-feature-management/latest/reporting/configuration-fetched-handler), flagd). This can allow us to use an event-based paradigm for consuming flags. Client apps may use feature flags for characteristics that aren't specifically tied to a user-action, making imperative flag evaluation a less-than-ideal solution. A server application could listen to a flag that changes some operational behavior.

Author: Added a blurb about this.

Examples:

- A web client's banner color could be updated when an associated flag is changed, without any user action or a page reload.

> - A service could subscribe to a flag that controls its global log-level.

Reviewer 8: oh yeah, I guess you're right. It's a bad example.

- A subsystem could be restarted with the new flag value.

## Design

> The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a set of events defined by the SDK (`ProviderEvents`). When a _application author_ registers a handler on a client, the client the client maintains this handler. The provider emits events or runs a callback indicating that it received a certain event, optionally providing data associated with that event. The callbacks registered with the client are then invoked with this data (`EventData`).

Author: Good point. I've added `application integrator`. Can you help me understand how an [integration author](https://docs.openfeature.dev/docs/specification/glossary#integration-author) (probably somebody who writes hooks) might use this API?

> The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a set of events defined by the SDK (`ProviderEvents`). When a _application author_ registers a handler on a client, the client the client maintains this handler. The provider emits events or runs a callback indicating that it received a certain event, optionally providing data associated with that event. The callbacks registered with the client are then invoked with this data (`EventData`).

Reviewer 8: The example that was top of mind for me was someone who's creating a react-specific feature flagging API that's built on top of Open Feature. I think that fits the persona of Integration Author? This is top of mind because I've been creating [exactly that API](https://github.com/moredip/openfeature-playground/blob/e41e472f45eff9b36a54ceb5ff66e8e1546c77c2/libs/openfeature-hooks/src/lib/useFlag.ts#L24) (a react hook for Open Feature), and need these events in order to make it work in an idiomatic react way.

> The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a set of events defined by the SDK (`ProviderEvents`). When a _application author_ registers a handler on a client, the client the client maintains this handler. The provider emits events or runs a callback indicating that it received a certain event, optionally providing data associated with that event. The callbacks registered with the client are then invoked with this data (`EventData`).

Author: You're absolutely right. I was being myopic thinking of hooks. Will add that persona.

> In the case of the aforementioned flag systems, flags are not actually evaluated when configurations are changed. One reason is that no dynamic context can be reasonably provided in the case of events, since the event is driven by a change in the flag management system, not a user-action. In these cases, the `EventData` would not contain flag values and the application author would have to evaluate flags in the registered handler. This is consistent with the event APIs already existing in those systems.

Reviewer: Counterpoint: in the mobile world, we don't evaluate on the client. "this thing changed" *is* accompanied with values, because we only dispatch on the things that we know about the user (device type and app version).

Author: This is a good point. I think that to maintain as similar an API as possible between client and server impls, it's ideal not to include flag values in the event. In the case of mobile web, as you point out, the flags are probably cached. As long as we consider race conditions, and define our ordering rigorously, I'd hope the cache of flags would be up to date, and subsequent evaluations after the events fire would return accurate flag values.

> In the case of the aforementioned flag systems, flags are not actually evaluated when configurations are changed. One reason is that no dynamic context can be reasonably provided in the case of events, since the event is driven by a change in the flag management system, not a user-action. In these cases, the `EventData` would not contain flag values and the application author would have to evaluate flags in the registered handler. This is consistent with the event APIs already existing in those systems.

Reviewer 8: FWIW, I agree with this approach of pushing responsibility for tracking what needs to be re-evaluated to the consumer.

How it looks implemented in a Provider:

```ts

> import { Provider, EventingProvider, ProviderEvents } from './types';

Author: I've updated the proposal concerning this ([here is the diff](https://github.com/open-feature/ofep/pull/25/commits/052ceb934830cdeb93042745851958240ab426a6)).

> Do we have a standardized set of events as part of the spec? Or do we keep to just standardizing a few key event names (that would be my preference) and then plan on standardizing other common event typer that might emerge from the community.

I'm interested in this approach, similar to what we have for `Reason`, we can have some predefined ones but allow any string to be used here.

> Also, do we want to support put anything in the spec about any sort of payload that comes with the event?

I've roughly defined the change data structure a bit.

> In your example you're implying that there would be a ConfigurationChanged event that would also include a payload describing flag data

I am proposing metadata, but not "data" (evaluated value). What's being (optionally) provided in the case of a change, is a list of flags changed (so they can be evaluated intelligently) and some free-form metadata (similar to the existing FlagMetadata proposal).

/**
 * The No-op provider is set by default, and simply always returns the default value.
 */
class MyEventingProvider implements Provider, EventingProvider {
  readonly metadata = {
    name: 'My Eventing Provider',
  } as const;

  // ...

  // the SDK listens for events, and fires associated handlers the application-author adds.
  readonly events = new EventEmitter();

  // pollDataSource a conceptual method specific to this provider that fires a callback if the flag source-of-truth of this provider changes.
  this.pollDataSource((newFlagData) => this.events.emit(ProviderEvents.ConfigurationChanged, newFlagData))

  // ...

```

How it looks for an `application author`:

```ts
OpenFeature.setProvider(new MyEventingProvider());

const client = OpenFeature.getClient();

// subscribe to ProviderEvents.ConfigurationChanged events

> client.addHandler(ProviderEvents.ConfigurationChanged, (eventData: EventData | undefined) => {

Reviewer: I think I'd like the ability to scope down this handler to only events of a particular key.
```java
logger.setLogLevel(client.getInteger('log_level', 4));
client.onChange('log_level', (i) -> logger.setLogLevel(i));
```

Author: I understand that, and it many cases it should be possible. It's not possible in all cases however, since there are SDKs which only send events indicating a general change occurred, not the flags changed.

In such cases, it could mean "run all handlers" or "run no handlers" or the provider might provide a configuration option in its constructor to choose between these behaviors.

Any takes on this?

Reviewer 7: > I understand that, and it many cases it should be possible. It's not possible in all cases however, since there are SDKs which only send events indicating a general change occurred, not the flags changed.

Isn't that where provider features could come into play?

Author: > Isn't that where provider features could come into play?

Provider features don't address the above questions totally. They will give the SDK an idea what this provider supports, but we still have the same fundamental question. Do we fire the handler or not? Do we make it configurable? The handler still has to do or not do _something_.

Reviewer 8: My 2 cents is that we should make implementation of an API like @justinabrahms describes a userland concern, built on top of this core capability, rather than something that's supported directly by the spec/SDKs

## 007-OFEP-flag-change-events

> ## State: DRAFTING

Author: ```suggestion
## State: APPROVED
```

Some flag SDKs support listening for flag value changes or general configuration changes ([launchdarkly](https://docs.launchdarkly.com/sdk/features/flag-changes), [cloudbees](https://docs.cloudbees.com/docs/cloudbees-feature-management/latest/reporting/configuration-fetched-handler), flagd). This can allow us to use an event-based paradigm for consuming flags. Client apps may use feature flags for characteristics that aren't specifically tied to a user-action, making imperative flag evaluation a less-than-ideal solution. Web clients using frameworks such as React need a way to register a callback for when a provider has initialized, given that it's not uncommon for a flag to be evaluated before the provider is ready (during page load). A server application could listen to a flag that changes some operational behavior.

Examples:

- A web client's banner color could be updated when an associated flag is changed, without any user action or a page reload.
- A service could subscribe to a flag that controls its global log-level.
- A subsystem could be restarted with the new flag value.

## Design

> The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a set of events defined by the SDK (`ProviderEvents`). When a consumer (_application integrator_, _application author_, _integration author_) registers a handler on a client, the client maintains this handler. The provider emits events or runs a callback indicating that it received a certain event, optionally providing data associated with that event. The callbacks registered with the client are then invoked with this data (`EventData`).

Reviewer: >the client maintains this handler

"client maintains" or "api maintains"? Which level of the hierarchy do these sit?

Author: I've proposed (and PoC'd) this functionality exists on the client.

> The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a set of events defined by the SDK (`ProviderEvents`). When a consumer (_application integrator_, _application author_, _integration author_) registers a handler on a client, the client maintains this handler. The provider emits events or runs a callback indicating that it received a certain event, optionally providing data associated with that event. The callbacks registered with the client are then invoked with this data (`EventData`).

Reviewer: ```suggestion
The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a set of events defined by the SDK (`ProviderEvents`). When a consumer (_application integrator_, _application author_, _integration author_) registers a handler on a client, the client maintains this handler. The provider emits events or runs a callback indicating something happened, optionally providing data associated with that event. The callbacks registered with the client are then invoked with this data (`EventData`).
```

emits an event when it receives an event sounds really weird.
